### PR TITLE
New features for future updates

### PR DESCRIPTION
**Pull request by MegadronA03**
### Summary:

Added:
 - Underwater geometry distortions to entity and banner models
 - Texture mapping code optimizations

Fixed:
 - Reflection mapping

TODO:
 - Fix entity fog and light
 - Add correct TBN for entity PBR and plants PBR

---------
### Screenshots:

Screenshot_20210718-022333_Minecraft
Screenshot_20210720-074712_Minecraft

### CODE REVIEW

File: shaders/glsl/includes/texture_mapping.glsl [hunk 1]
```diff
@@ -9,49 +9,39 @@
...
+
             if(baseNormal.g > 0.9){
                 reliefMap.gb = reliefMap.bg;
-                reliefMap = reliefMap * 2.0 - 1.0;
                 reliefMap.rb *= normalMapStrength;
                 baseNormal = normalize(reliefMap);
             }else{
                 if(baseNormal.g < -0.9){
-                    reliefMap.b = -reliefMap.b;
+                    reliefMap.b *= -1.0;
```
Comment (jebbyk) [inline shaders/glsl/includes/texture_mapping.glsl:20]: I think simple sign changing is more optimized than multiplication

File: shaders/glsl/includes/texture_mapping.glsl [hunk 2]
```diff
@@ -9,49 +9,39 @@
...
                     reliefMap.gb = reliefMap.bg;
-                    reliefMap = reliefMap * 2.0 - 1.0;
                     reliefMap.rb *= normalMapStrength;
                     baseNormal = normalize(reliefMap);
                 }else{
                     if (baseNormal.b > 0.9){
-                        reliefMap.g = 1.0 - reliefMap.g;// OpenGl needs G to be flipped
-                        reliefMap = reliefMap * 2.0 - 1.0;
+                        reliefMap.g *=  -1.0;
```
Comment (jebbyk) [inline shaders/glsl/includes/texture_mapping.glsl:26]: dont think it is correct. In this case .g will never be grater then 0

File: shaders/glsl/includes/texture_mapping.glsl [hunk 3]
```diff
@@ -155,11 +145,11 @@
                         reliefMap = mapWaterNormals(texture0);
                     #endif
                 }else{
-                    highp vec2 reliefMapCoord = diffuseMapCoord - vec2(0.0,  1.0 / (TEXTURE_ATLAS_DIMENSION.x * 2.0));
+                    highp vec2 reliefMapCoord = diffuseMapCoord - vec2(0.0, invTexDimD.y); //x coordinate had no sense, and after tex atlas in 1.17 were updated, x doubled relative to y, causing wrong texture reading and also there must be initially y because of vec2
```
Comment (jebbyk) [inline shaders/glsl/includes/texture_mapping.glsl:148]: hehe buugzzz :-)

File: shaders/glsl/renderchunk.vertex [hunk 4]
```diff
@@ -165,17 +166,9 @@ POS4 worldPos;
...
-if(isUnderWater > 0.9){
-	highp float t = TIME;
-	float wScale = 0.0125 * length(worldPos.xyz);
-	float wMovingSpeed = 4.0;
-	float wFreq  = 32.0 / length(worldPos.xyz);
-
-	gl_Position.x += (1.0 + sin(t * wMovingSpeed + (worldPos.x + worldPos.z + worldPos.y) * wFreq)) * wScale;  
-}
+
+bool isUWbool = (detectUnderwater() > 0.9);
```
Comment (jebbyk) [inline shaders/glsl/renderchunk.vertex]: I think _isUnderWater_ is more understandable one

**Review by jebbyk — COMMENTED** [review]
I haven't tested it on my phone tho

File: shaders/glsl/includes/texture_mapping.glsl (hunk 2 above)
Comment (MegadronA03) [reply shaders/glsl/includes/texture_mapping.glsl:26]: I already converted it at beginning (* 2.0 - 1.0) I tested it results are same and flipping g is same as multiplication to -1 but with offset
Screenshot_20210722-141020_Desmos

File: shaders/glsl/includes/texture_mapping.glsl (hunk 1 above)
Comment (MegadronA03) [reply shaders/glsl/includes/texture_mapping.glsl:20]: Yeah, forgot about that

Comment (jebbyk) [reply shaders/glsl/includes/texture_mapping.glsl:20]: ok, Ill check it

**Comment by OpenSauce04** [comment]
Are you sure that this isn't going to completely break the shader when it's pulled from how old it is monkaGun

**Comment by jebbyk** [comment]
I think he has checked it on his own phone. And also I will check it too when I have time. So nothing to worry about

**Pull request merged**